Fix song select debounce not handling long (stutter) frames well

peppy opened the pull request:
This fixes the edge case where a long single frame during key repeat iteration causes debounce to run immediately, instead of an expectation that it would just ignore the stutter frame and carry on.

Without this, we [runr into scenarios](https://discord.com/channels/188630481301012481/1097318920991559880/1412657191780880526) where one selection change during iteration cascades into a very bad user experience (first debounce being performed leads to further stutters, feedback loop etc.)

For what it's worth, I can reproduce this on my debug builds due to not having satori running (and something on my system causing stutters for whatever reason.

- I've also documented the debounces better, and adjusted up slightly.
- I tested adding debounce to the beatmap metadata fetch (as we're doing a *lot* of requests right now) but it doesn't work well due to being a debounce-reliant-on-another-debounce. A better approach would likely be to back-off when many selections happen over a short period, rather than delaying every request. Something for future exploration.

peppy committed:
Implement local song select debounce rather than using `Scheduler.AddDelayed`

This is to allow more custom handling of the debounce timing.

peppy committed:
Fix song select debounce happening too early during fast iteration if there's a stutter

peppy committed:
Adjust song select debounce upwards slightly

I still think it can probably go higher. Just going to do this in small
increments until people complain / notice.

peppy committed:
Better document various debounce constants and split out processing debounce for clarity

bdach committed:
Merge branch 'master' into song-select-better-debounce

bdach commented on osu.Game/Screens/SelectV2/SongSelect.cs:
```diff
@@ -63,9 +63,20 @@ namespace osu.Game.Screens.SelectV2
+        /// either after selection or after a panel comes on screen. Value should be low enough that users don't complain,
+        /// but otherwise as high as possible to reduce overheads.
+        /// </summary>
+        public const int DIFFICULTY_CALCULATION_DEBOUNCE = 150;
```
Is there a reason why this is a separate constant set to the same value as `SELECTION_DEBOUNCE`?

bdach commented on osu.Game/Screens/SelectV2/SongSelect.cs:
```diff
@@ -371,8 +382,57 @@ protected override void Update()
+        {
+            if (debounceQueuedSelection == null) return;
+
+            // avoid debounce running early if there's a single long frame.
+            debounceElapsedTime += Math.Min(1000 / 60.0, Clock.ElapsedFrameTime);
```
Not sure how I feel about this. Maybe it works for a single long frame, but if someone is, say, only able to pull 30fps in song select this will effectively double the debounce duration. Thinking about especially about mobile performance here.

peppy replied:
valid concern, i'll adjust.

bdach commented:
Please double-check merge conflict resolution as well.

peppy committed:
Change debounce consideration to use dynamic FPS moving average rather than fixed 60 fps

This should better account for scenarios where user FPS is below 60 fps.
Previously the debounce would unexpectedly be longer than usual for low
FPS scenarios.

bdach commented on osu.Game/Screens/SelectV2/SongSelect.cs:
```diff
@@ -371,8 +382,57 @@ protected override void Update()
+            if (debounceQueuedSelection == null) return;
+
+            // avoid debounce running early if there's a single long frame.
+            debounceElapsedTime += Math.Min(1000 / Clock.FramesPerSecond, Clock.ElapsedFrameTime);
```
I'm choosing to ignore the fact that many places in framework choose to leave `Clock.FramesPerSecond` unimplemented by returning 0 (or return 0 as a sort of initial state for the clock), because the semantics of floating point are *for once* nice enough anyway such that the `Math.Min()` call just reduces to returning `Clock.ElapsedFrameTime` in such a case.

bdach reviewed (dismissed):
let's see how this goes I guess

bdach commented:
That said the test failures here are still suspicious. Pretty sure I've never seen some of these tests fail until now.

peppy committed:
Fix multiple concerns with new debounce logic

Tried many approaches but this seems simplest to guarantee no test (or
other) regressions..

peppy commented:
I've pushed a fix for all the issues I could find. Not sure if you'll agree with it.

bdach commented on osu.Game/Screens/SelectV2/SongSelect.cs:
```diff
@@ -383,7 +384,8 @@ protected override void Update()
             };
 
-            updateDebounce();
+            if (this.IsCurrentScreen())
```
What's this about? I don't like `TestFilteringRunsAfterReturningFromGameplay` failing just after this gets added. I'm not sure I recall it ever failing before. I'm not sure it's related, but it might be?

peppy replied:
I added this to fix tests, figured something something different order of execution (try removing, will get assertion failures on disabled bindable operations, 100% repro).

> I don't like TestFilteringRunsAfterReturningFromGameplay failing just after this gets added.

I think it's failing on other branches too 😦 tried to fix last night with a [commit to master](https://github.com/peppy/osu/commit/71d0afd4c2b3c7fe6e92558b5ecd849f67cb6384) but it's failing in a different place to expected..

bdach replied:
Well if the test is failing somewhere else already then I guess it's probably fine......

I guess next immediate order of business for me would be to go fix flaky tests on master because we're at the stage where it's unclear whether 4 reds is signal or noise.
